ledger/src/shred.rs

```diff
@@ -467,8 +473,36 @@ impl Shred {
         self.common_header.signature
     }
 
-    pub fn seed(&self) -> [u8; 32] {
+    pub fn seed(
+        &self,
+        leader_schedule_cache: &Arc<LeaderScheduleCache>,
+        bank_forks: &Arc<RwLock<BankForks>>,
+    ) -> [u8; 32] {
         let mut seed = [0; 32];
+
+        let bf = bank_forks.read().unwrap();
+        let maybe_bank = bf.get(self.slot());
+        if let Some(bank) = maybe_bank {
+            if enable_deterministic_seed(self.slot(), bank) {
+                let maybe_leader_pubkey =
+                    leader_schedule_cache.slot_leader_at(self.slot(), Some(bank));
```

Comment:
Here I am not sure if we should use the root bank, working bank, or the bank for the slot. Maybe @carllin can confirm.

In `retransmit_stage.rs` seems like we are using the working-bank, which now I am thinking maybe it is not correct:
https://github.com/solana-labs/solana/blob/db3475bcd/core/src/retransmit_stage.rs#L412

Reply:
>In retransmit_stage.rs seems like we are using the working-bank, >which now I am thinking maybe it is not correct:

hmmm yeah, I think what's important there is not so much which bank we use, but which epoch we're sampling the stakes from. Ideally, for a particular packet `P`, we would want people to use the stakes from the same epoch when performing the shuffle.

 The issue with that code currently is it's using the working bank will probably not be consistent at epoch boundaries, where some nodes will have a later working bank than others. Using the rooted bank's epoch runs into the same problem, some nodes may have earlier/later rooted banks than others that either don't cross/cross an epoch boundary. 

Ideally what would be done there I think is to use the epoch stakes from the epoch for which the slot of packet `P` belongs, which is consistently observable across all nodes. The implementation probably didn't do this because the packet `P` doesn't have easy access to that slot information, it would need to deserialize that information from the packet data, based on the shred offset of the Slot in the shred headers.

> Here I am not sure if we should use the root bank, working bank, or the bank for the slot.

Here it should be fine to use the bank for the shred `slot`, since the leader schedule for this epoch was determined 2 epochs ago, so it should be consistent across all nodes across all slots in the current epoch.

Reply:
@carllin can we use the root bank here for `slot_leader_at`?

Reply:
> @carllin can we use the root bank here for `slot_leader_at`?
> 
> I think the code here will be simpler and more efficient if we can just use the root-bank for `slot_leader_at`.

Yup! It's fine to use the root bank here instead of the `slot_bank`, rendering the `maybe_slot_bank` unnecessary and saving a lookup. Good call 😃

Reply:
Thanks! updated